src/tequila/circuit/circuit.py

```diff
@@ -521,6 +524,95 @@ def map_qubits(self, qubit_map):
         # currently its recreated in the init function
         return QCircuit(gates=new_gates)
 
+    def add_controls(self, control, inpl: typing.Optional[bool] = True) \
+            -> typing.Optional[QCircuit]:
+        """Depending on the truth value of inpl:
+            - return controlled version of self with control as the control qubits if inpl;
+            - mutate self so that the qubits in control are added as the control qubits if not inpl.
+
+        Raise ValueError if there any qubits in common between self and control.
+        """
+        if inpl:
+            self._inpl_control_circ(control)
+        else:
+            return self._mutate_control_circ(control)
```

Comment:
could this be replaced by:

U = copy.deepcopy(self)
return U.add_controls(..., inpl=True)

?